### PR TITLE
New feature uploader

### PR DESCRIPTION
**Pull request by arejula27**
I have created a new endpoint that allows the uploading of a CSV file. This file will be stored on Hugging Face and linked to the dataset using its identifier.

### CODE REVIEW

File: src/uploader/hugging_face_uploader.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                    path_in_repo=f"/data/{file.filename}",
+                    repo_id=repo_id,
+                    repo_type="dataset",
+                    token=token,
+                )
+
+            except Exception:
+                huggingface_hub.delete_repo(repo_id, token=token, repo_type="dataset")
+                msg = "Error uploading the file"
+                raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=msg)
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:54]: 502 error would be more appropriate, I think (same with the upload_file error)

File: src/uploader/hugging_face_uploader.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            try:
+                huggingface_hub.upload_file(
+                    path_or_fileobj=self._generate_metadata_file(dataset),
+                    path_in_repo="README.md",
+                    repo_id=repo_id,
+                    repo_type="dataset",
+                    token=token,
+                )
+
+            except Exception as e:
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py]: You're catching any error in `_generate_metadata_file` here, and stating that it's an error in huggingface. I think you should have the `_generate_metdata_file` outside this try

File: src/uploader/hugging_face_uploader.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _generate_metadata_file(self, dataset: Dataset) -> bytes:
+        tags = ["- " + tag.name for tag in dataset.keywords] if dataset.keywords else []
+        content = "---\n"
+        content += f"pretty_name: {dataset.name}\n"
+
+        if tags:
+            content += "tags:\n"
+            content += "\n".join(tags) + "\n"
+        # TODO the license must be in the hugginface format:
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:116]: I think that licenses are very important for datasets. It would therefore be good to fix this TODO now, I think.
We can for now raise an error if the license is not supported by HuggingFace (because we don't have a AIoD standard for now). Once we have an AIoD standard, we might want to make a translation.

File: src/uploader/hugging_face_uploader.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        #  https://huggingface.co/docs/hub/repositories-licenses
+        """
+        if dataset.license:
+            content += f"license: {dataset.license.name if dataset.license else ''}"
+        """
+        content += "---\n"
+        content += f"# {dataset.name}\n"
+        content += "Created from AIOD platform"  # TODO add url
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:124]: Is this TODO for this PR?

I think it would be nice to add some additional fields. https://huggingface.co/datasets/openwebtext/raw/main/README.md is an example. It would be nice to `dataset_info` metadata, authors, contacts, citation, ....

File: src/main.py [hunk 5]
```diff
@@ -150,6 +151,8 @@ def get_platforms() -> list:
     for router in routers.routers:
         app.include_router(router.create(engine, url_prefix))
 
+    app.include_router(HuggingfaceUploader().create(engine))
```
Comment (josvandervelde) [inline src/main.py:154]: Adding `url_prefix` would be good. This makes it possible to set the url to `test.openml.org/demo/upload/...` ("demo" is then the url_prefix)

File: src/uploader/hugging_face_uploader.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        )
+        return router
+
+    def _handle_upload_func(
+        self,
+        engine: Engine,
+    ):
+        def handle_upload(
+            resource: int,
+            file: UploadFile = File(...),
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:31]: Why the `=File(...)`, and the `=Form(...)`s? What does it do? Why the ellipsis (...)?
I'm unfamiliar with this usage, and I'm not sure why we need default values for these fields

File: src/uploader/hugging_face_uploader.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                    repo_type="dataset",
+                    token=token,
+                )
+
+            except Exception:
+                huggingface_hub.delete_repo(repo_id, token=token, repo_type="dataset")
+                msg = "Error updating metadata in huggingface"
+                raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=msg)
+
+            # TODO create and save  DataDownload within the dataset
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py]: As I understand, this TODO will be fixed in this PR, right?

File: src/uploader/hugging_face_uploader.py (hunk 4 above)
Comment (arejula27) [reply src/uploader/hugging_face_uploader.py:114]: Agree, but i prefer to deliver a small functionality and give more time to martin

File: src/uploader/hugging_face_uploader.py (hunk 6 above)
Comment (arejula27) [reply src/uploader/hugging_face_uploader.py]: Is how i found in internet it must be done, if there is a better approach tell me and i change it

**Comment by arejula27** [comment]
Sometimes error is throw 
```bash
sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <Dataset at 0xffff743422d0> is not bound 
to a Session; lazy load operation of attribute 'keywords' cannot proceed 
(Background on this error at: https://sqlalche.me/e/14/bhk3)
```

**Comment by arejula27** [comment]
All changes made

File: src/main.py [hunk 8]
```diff
@@ -150,6 +151,15 @@ def get_platforms() -> list:
     for router in routers.routers:
         app.include_router(router.create(engine, url_prefix))
 
+    @app.post(url_prefix + "/upload/datasets/{identifier}/huggingface")
+    def hugginffaceUpload(
+        identifier: int,
+        file: UploadFile,
+        token: str,
```
Comment (josvandervelde) [inline src/main.py:158]: Some documentation would be nice, I think. What kind of token is this? What kind of username? It would be nice if that shows up in the swagger

File: src/uploader/hugging_face_uploader.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                session.merge(resource)
+                session.commit()
+            except Exception as e:
+                raise HTTPException(
+                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="TODO: describe"
+                ) from e
+
+    def _create_or_get_repo_url(sef, repo_id, token):
+        try:
+            datasets.load_dataset_builder(repo_id)
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:93]: In my tests, I thought that this didn't work. I thought that this will not give a response when the repo exists, but is empty. In that case, the create_repo will throw an error, if I'm correct.

File: src/uploader/hugging_face_uploader.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    def _store_resource_updated(self, engine: Engine, resource: Dataset, url: str, repo_id: str):
+        with Session(engine) as session:
+            try:
+                data_download = DataDownloadORM(content_url=url, name=repo_id, dataset=resource)
+                resource.distributions.append(data_download)
+                session.merge(resource)
+                session.commit()
+            except Exception as e:
+                raise HTTPException(
+                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="TODO: describe"
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:88]: There is a TODO here

File: src/tests/uploader/huggingface/test_dataset_uploader.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            params={"username": "Fake-username", "token": "Fake-token"},
+            headers={"Authorization": "Fake token"},
+            files=files,
+        )
+    assert response.status_code == 200
+    id_response = response.json()
+    assert id_response == dataset_id
+
+
+# TODO: add test for when respository exists already, but is empty (I don't think the current
```
Comment (josvandervelde) [inline src/tests/uploader/huggingface/test_dataset_uploader.py:64]: There are some TODOs here. I think that some use-cases don't work right now, and that the error handling is not great. So a couple of tests would really help, I think.

**Review by josvandervelde — COMMENTED** [review]
There were 6 issues still open. They were hidden, so you probably missed them. Could you also take a look at those?

File: src/uploader/hugging_face_uploader.py (hunk 9 above)
Comment (arejula27) [reply src/uploader/hugging_face_uploader.py:108]: I dont understand

File: src/uploader/hugging_face_uploader.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                token=token,
+            )
+        except HTTPError as e:
+            msg = f"Error uploading the file, huggingface api returned a http error: {e.strerror}"
+            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=msg)
+
+        except ValueError:
+            msg = "Error uploading the file, bad format"
+            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=msg)
+        # RepositoryNotFoundError and RevisionNotFoundError can not be imported
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:45]: Why not? `from huggingface_hub.utils import RepositoryNotFoundError` ?

**Comment by arejula27** [comment]
the option to attach a document just disappeared from swager

File: src/uploader/hugging_face_uploader.py (hunk 9 above)
Comment (josvandervelde) [reply src/uploader/hugging_face_uploader.py:120]: So you're depending on the behaviour of `datasets.load_dataset_builder`
You want that it throws an error if and only if the repository does not exist.
I thought I saw different behaviour, I thought that it also throws an error if there is a repo but no files yet.

Anyway, I think that a bit more explicit would be to catch a `RepositoryNotFoundError`. I'm not sure if this `load_dataset_builder` throws it, but `model_info` seems to do that. I'm not saying you should use that one, I'm just saying I'm not sure this will work.

File: src/tests/uploader/huggingface/test_dataset_uploader.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (josvandervelde) [inline src/tests/uploader/huggingface/test_dataset_uploader.py:64]: Can these TODOs be deleted?

File: src/tests/uploader/huggingface/test_dataset_uploader.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            params={"username": "Fake-username", "token": "Fake-token"},
+            headers={"Authorization": "Fake token"},
+            files=files,
+        )
+    assert response.status_code == 200
+    id_response = response.json()
+    assert id_response == dataset_id
+
+
+# TODO: tests some error handling?
```
Comment (josvandervelde) [inline src/tests/uploader/huggingface/test_dataset_uploader.py:118]: Should this TODO be removed? I guess you don't want to make them?

File: src/uploader/hugging_face_uploader.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+        try:
+            huggingface_hub.upload_file(
+                path_or_fileobj=io.BufferedReader(file.file),
+                path_in_repo=f"/data/{file.filename}",
+                repo_id=repo_id,
+                repo_type="dataset",
+                token=token,
+            )
+        except HTTPError as e:
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:38]: A bit of a nit, ignore this if you want. But:

You've got a lot of code duplication in the error handling now. 
It would be nice to remove the code duplication. Also I'm not sure if you really need to catch the "repository does not exist" and "revision does not exist" separate from the "unexpected error". Those are also unexpected, I'd say. They should never happen, because you're creating the repo before. I think you could remove those separate catches.

File: src/uploader/hugging_face_uploader.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            msg = f"Error uploading the file, huggingface api returned a http error: {e.strerror}"
+            raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=msg)
+
+        except ValueError:
+            msg = "Error uploading the file, bad format"
```
Comment (josvandervelde) [inline src/uploader/hugging_face_uploader.py:43]: You're now completely ignoring the error message. I think that it would be nice to give some error message back to the user in all error messages, similar as you do for (some) of the other errors.

There might be information in the HuggingFace error message (field "foo" is required) and that's completely ignored now.

File: src/uploader/hugging_face_uploader.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _get_resource(self, engine: Engine, identifier: int) -> Dataset:
+        """
+        Get the resource identified by AIoD identifier
+        """
+
+        with Session(engine) as session:
+            query = (
+                session.query(Dataset)
+                .options(lazyload("*"))
```
[the remark this answers is not in the document]
Comment (arejula27) [reply src/uploader/hugging_face_uploader.py]: This has change, i load only the needed relations, only one session is used

File: src/tests/uploader/huggingface/test_dataset_uploader.py (hunk 13 above)
Comment (arejula27) [reply src/tests/uploader/huggingface/test_dataset_uploader.py:64]: sure

File: src/tests/uploader/huggingface/test_dataset_uploader.py (hunk 14 above)
Comment (arejula27) [reply src/tests/uploader/huggingface/test_dataset_uploader.py:118]: I think that this todo can remain, dont need to be done on this pr

File: src/uploader/hugging_face_uploader.py (hunk 15 above)
Comment (arejula27) [reply src/uploader/hugging_face_uploader.py:38]: I created cuz your previous comment 🙃

**Review by josvandervelde — APPROVED** [review]
